src/mscorlib/src/System/ArraySegment.cs

```diff
@@ -110,6 +110,87 @@ public int Count
             }
         }
 
+        public int Length => Count;
+
+        public T this[int index]
+        {
+            get
+            {
+                if (_array == null)
+                    ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_NullArray);
+                if (index < 0 || index >= _count)
+                    ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+                Contract.EndContractBlock();
+
+                return _array[_offset + index];
+            }
+
+            set
+            {
+                if (_array == null)
+                    ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_NullArray);
+                if (index < 0 || index >= _count)
+                    ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+                Contract.EndContractBlock();
+
+                _array[_offset + index] = value;
+            }
+        }
+
+        /// <summary>
+        /// Returns an empty <see cref="ArraySegment{T}"/>
+        /// </summary>
+        public static ArraySegment<T> Empty
+        {
+            get { return default(ArraySegment<T>); }
+        }
+
+        public ArraySegment<T> Slice(int start) => Slice(start, _count - start);
+
+        public ArraySegment<T> Slice(int start, int length)
+        {
+            if (_array == null)
+                ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_NullArray);
+            if ((uint)start > (uint)_count || (uint)length > (uint)(_count - start))
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+            Contract.EndContractBlock();
+
+            return new ArraySegment<T>(_array, _offset + start, length);
+        }
+
+        public void CopyTo(T[] destination) => CopyTo(destination, 0);
+
+        public void CopyTo(T[] destination, int destinationIndex)
```

Comment:
Need to validate that `destinationIndex` is a valid index for destination.

If `Array.Copy` will do the validation, please add a comment to that effect.